### ISSUE TITLE
Error linking: missing OpenSSL TLS1.1/1.2 symbols

### ISSUE THREAD

**Issue by umareddy** [facebook/osquery#1373]
I get the following undefined symbol error, looks like I am not linking against the correct openssl library. I did install openssl using "brew install openssl"

```
[ 89%] Built target osquery_dispatcher
[ 89%] Built target osquery_distributed
[ 89%] Built target osquery_events
[ 89%] Built target libosquery
Linking CXX executable osqueryd
Undefined symbols for architecture x86_64:
  "_ERR_remove_thread_state", referenced from:
      boost::asio::ssl::detail::openssl_init_base::do_init::~do_init() in libosquery_additional.a(tls.cpp.o)
      boost::asio::ssl::detail::openssl_init_base::do_init::~do_init() in libosquery_additional.a(tls.cpp.o)
      boost::asio::ssl::detail::openssl_init_base::do_init::~do_init() in libosquery_additional.a(tls.cpp.o)
      boost::asio::ssl::detail::openssl_init_base::do_init::~do_init() in libosquery_additional.a(tls.cpp.o)
  "_TLSv1_1_client_method", referenced from:
      boost::asio::ssl::context::context(boost::asio::ssl::context_base::method) in libcppnetlib-client-connections.a(client.cpp.o)
  "_TLSv1_1_method", referenced from:
      boost::asio::ssl::context::context(boost::asio::ssl::context_base::method) in libcppnetlib-client-connections.a(client.cpp.o)
  "_TLSv1_1_server_method", referenced from:
      boost::asio::ssl::context::context(boost::asio::ssl::context_base::method) in libcppnetlib-client-connections.a(client.cpp.o)
  "_TLSv1_2_client_method", referenced from:
      boost::asio::ssl::context::context(boost::asio::ssl::context_base::method) in libcppnetlib-client-connections.a(client.cpp.o)
  "_TLSv1_2_method", referenced from:
      boost::asio::ssl::context::context(boost::asio::ssl::context_base::method) in libcppnetlib-client-connections.a(client.cpp.o)
  "_TLSv1_2_server_method", referenced from:
      boost::asio::ssl::context::context(boost::asio::ssl::context_base::method) in libcppnetlib-client-connections.a(client.cpp.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [osquery/osqueryd] Error 1
make[2]: *** [osquery/CMakeFiles/daemon.dir/all] Error 2
make[1]: *** [all] Error 2
make: *** [all] Error 2
```

**Comment by theopolis** [comment]
Hey @umareddy! Did you install osquery using brew? If so can you do a `brew info osquery`?

On my system I have osquery installed via the package on https://osquery.io, which links against the Apple-provided OpenSSL:

```
/usr/local/bin/osqueryi:
	/usr/lib/libcrypto.0.9.8.dylib (compatibility version 0.9.8, current version 0.9.8)
	/usr/lib/libssl.0.9.8.dylib (compatibility version 0.9.8, current version 0.9.8)
	/usr/lib/libedit.3.dylib (compatibility version 2.0.0, current version 3.0.0)
	/usr/lib/libncurses.5.4.dylib (compatibility version 5.4.0, current version 5.4.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1213.0.0)
	/usr/lib/libbz2.1.0.dylib (compatibility version 1.0.0, current version 1.0.5)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.5)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 1151.16.0)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 62.0.0)
	/System/Library/Frameworks/SystemConfiguration.framework/Versions/A/SystemConfiguration (compatibility version 1.0.0, current version 699.1.5)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/DiskArbitration.framework/Versions/A/DiskArbitration (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1151.16.0)
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 57031.1.35)
	/System/Library/Frameworks/OpenDirectory.framework/Versions/A/OpenDirectory (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.0.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
```

When osquery is built as a brew bottle it will compile and link against the brew-provided OpenSSL. Those missing symbols for TLS1.1 and TLS1.2 indicate your osquery binary is definitely looking for brew's OpenSSL as Apple's 0.9.8 only includes support for TLS1.0. :/

**Comment by theopolis** [comment]
Just a bit more debugging:

```
$ brew install osquery
==> Installing osquery dependency: openssl
==> Downloading https://homebrew.bintray.com/bottles/openssl-1.0.2d_1.yosemite.bottle.tar.gz
######################################################################## 100.0%
==> Pouring openssl-1.0.2d_1.yosemite.bottle.tar.gz
==> Caveats
A CA file has been bootstrapped using certificates from the system
keychain. To add additional certificates, place .pem files in
  /usr/local/etc/openssl/certs

and run
  /usr/local/opt/openssl/bin/c_rehash

This formula is keg-only, which means it was not symlinked into /usr/local.

OS X already provides this software and installing another version in
parallel can cause all kinds of trouble.

Apple has deprecated use of OpenSSL in favor of its own TLS and crypto libraries

Generally there are no consequences of this for you. If you build your
own software and it requires this formula, you'll need to add to your
build variables:

    LDFLAGS:  -L/usr/local/opt/openssl/lib
    CPPFLAGS: -I/usr/local/opt/openssl/include

==> Summary
🍺  /usr/local/Cellar/openssl/1.0.2d_1: 464 files, 18M
==> Installing osquery
==> Downloading https://homebrew.bintray.com/bottles/osquery-1.5.0_1.yosemite.bottle.tar.gz
######################################################################## 100.0%
==> Pouring osquery-1.5.0_1.yosemite.bottle.tar.gz
Error: The `brew link` step did not complete successfully
The formula built, but is not symlinked into /usr/local
Could not symlink bin/osqueryctl
Target /usr/local/bin/osqueryctl
already exists. You may want to remove it:
  rm '/usr/local/bin/osqueryctl'

To force the link and overwrite all conflicting files:
  brew link --overwrite osquery

To list all files that would be deleted:
  brew link --overwrite --dry-run osquery

Possible conflicting files are:
/usr/local/bin/osqueryctl
/usr/local/bin/osqueryd
/usr/local/bin/osqueryi
/usr/local/include/osquery/config.h
/usr/local/include/osquery/core.h
/usr/local/include/osquery/database.h
/usr/local/include/osquery/enroll.h
/usr/local/include/osquery/events.h
/usr/local/include/osquery/extensions.h
/usr/local/include/osquery/filesystem.h
/usr/local/include/osquery/flags.h
/usr/local/include/osquery/hash.h
/usr/local/include/osquery/logger.h
/usr/local/include/osquery/registry.h
/usr/local/include/osquery/sdk.h
/usr/local/include/osquery/sql.h
/usr/local/include/osquery/status.h
/usr/local/include/osquery/tables.h
/usr/local/lib/libosquery.a
==> Summary
🍺  /usr/local/Cellar/osquery/1.5.0_1: 28 files, 22M
------------------------------------------------------------
$ otool -L /usr/local/Cellar/osquery/1.5.0_1/bin/osqueryi
/usr/local/Cellar/osquery/1.5.0_1/bin/osqueryi:
	/usr/local/lib/libgflags.2.dylib (compatibility version 2.0.0, current version 2.1.2)
	/usr/local/opt/openssl/lib/libcrypto.1.0.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/local/opt/openssl/lib/libssl.1.0.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libedit.3.dylib (compatibility version 2.0.0, current version 3.0.0)
	/usr/lib/libncurses.5.4.dylib (compatibility version 5.4.0, current version 5.4.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1213.0.0)
	/usr/lib/libbz2.1.0.dylib (compatibility version 1.0.0, current version 1.0.5)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.5)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 1153.20.0)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 62.0.0)
	/System/Library/Frameworks/SystemConfiguration.framework/Versions/A/SystemConfiguration (compatibility version 1.0.0, current version 699.30.1)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/DiskArbitration.framework/Versions/A/DiskArbitration (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1153.18.0)
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 57031.30.12)
	/System/Library/Frameworks/OpenDirectory.framework/Versions/A/OpenDirectory (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.0.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
------------------------------------------------------------
$ ls -la /usr/local/opt/openssl/lib
total 12528
drwxr-xr-x  10 reed  admin      340 Jul 13 11:46 .
drwxr-xr-x  11 reed  admin      374 Jul 22 14:17 ..
drwxr-xr-x  14 reed  admin      476 Jul 13 11:46 engines
-r--r--r--   1 reed  admin  1990028 Jul 22 14:16 libcrypto.1.0.0.dylib
-r--r--r--   1 reed  admin  3481904 Jul 13 11:46 libcrypto.a
lrwxr-xr-x   1 reed  admin       21 Jul 13 11:46 libcrypto.dylib -> libcrypto.1.0.0.dylib
-r--r--r--   1 reed  admin   367912 Jul 22 14:16 libssl.1.0.0.dylib
-r--r--r--   1 reed  admin   559048 Jul 13 11:46 libssl.a
lrwxr-xr-x   1 reed  admin       18 Jul 13 11:46 libssl.dylib -> libssl.1.0.0.dylib
drwxr-xr-x   5 reed  admin      170 Jul 22 14:17 pkgconfig
------------------------------------------------------------
$ /usr/local/Cellar/osquery/1.5.0_1/bin/osqueryi                                                                                                                                                                                                                                                                                                                    reed@reed-mbp
osquery - being built, with love, at Facebook
Using a virtual database. Need help, type '.help'
osquery>

**Comment by umareddy** [comment]
I did a brew install osquery
Installed without error, and I am able to run the command:
/usr/local/Cellar/osquery/1.5.0_1/bin/osqueryi

So I assume brew installed osquery is linking with the correct openssl library.

My make continues to fail with the same error.

Based on the following:
> When osquery is built as a brew bottle it will compile and link against the brew-provided OpenSSL. Those missing symbols for TLS1.1 and TLS1.2 indicate your osquery binary is definitely looking for brew's OpenSSL as Apple's 0.9.8 only includes support for TLS1.0. :/
> 
> 

I assume that I have to somehow link with the right openssl library. How do I do that? Do I have to change makefile and change the LD_LIBRARY_PATH? I am following the instructions provided here to clone from github and build.  https://osquery.io/downloads/ <https://osquery.io/downloads/>

Thanks for your help.

**Comment by theopolis** [comment]
Ah, perhaps your brew OpenSSL has installed it's headers into a path CMake is picking up? Maybe `/usr/local/include/openssl/`? If that path exists (or CMake is finding brew's newer OpenSSL headers) then the TLS1.1+ symbols/definitions will exist. This will cause osquery and cpp-netlib to include support for TLS1.1/TLS1.2, aka, use the respective methods in their client code. At link time, if you're linking using the Apple-provided OpenSSL these symbols wont exist.

**Comment by umareddy** [comment]
I am able to compile clean now. My problems were:
1. Including openssl header files from /usr/include/openssl
2. Linking with /usr/lib/libssl.0.9.8.dylib and /usr/lib/libcrypto.0.9.8.dylib

(These are provided by OS X, and Mac does not use the openssl libraries installed by brew)

I had a version (OpenSSL 1.0.1e 11 Feb 2013) of openssl installed under /opt/local 

( I did a 'brew install openssl’ which installed openssl version - OpenSSL 1.0.2d 9 Jul 2015 in /usr/local/Cellar/openssl I did not bother using this version for now)

I updated the osquery/build/darwin/CMakeCache.txt to change the path to find the headers and libraries. Here are the 3 lines I changed:
/Path to a library.
OPENSSL_CRYPTO_LIBRARY:FILEPATH=/opt/local/lib/libcrypto.dylib

//Path to a file.
OPENSSL_INCLUDE_DIR:PATH=/opt/local/include

//Path to a library.
OPENSSL_SSL_LIBRARY:FILEPATH=/opt/local/lib/libssl.dylib

Note that I changed /usr to /opt/local

I did:
make clean
make -j 8

Built clean.

Thanks for you help.

**Comment by theopolis** [comment]
Curious, is `/usr/include/openssl` a symlink to `/usr/local/Cellar/openssl/<version>/include/openssl`?

**Comment by umareddy** [comment]
/usr/include/openssl is NOT a symlink on my MBP

It would be perfect if /usr/include/openssl was a symlink. Not just headers but also 
/usr/lib/libssl.dylib
/usr/lib/libcrypto.dylib
/usr/bin/openssl

must be symlinked to the respective dir/files in /usr/local/Cellar/openssl/<version>

I figured OS X has some reason to continue to use their own ssl and crypto library.

The altervative I figured would work best on os x is to symlink:
/opt/local/include/openssl
/opt/local/lib/libssl.dylib
/opt/local/lib/libcrypto.dylib

to the respective dir/files in /usr/local/Cellar/openssl/<version>

AND make the following changes to CMakeCache.txt file
/Path to a library.
OPENSSL_CRYPTO_LIBRARY:FILEPATH=/opt/local/lib/libcrypto.dylib

//Path to a file.
OPENSSL_INCLUDE_DIR:PATH=/opt/local/include

//Path to a library.
OPENSSL_SSL_LIBRARY:FILEPATH=/opt/local/lib/libssl.dylib

I do not understand osx deep enough, do not know what I would break if I symlinked /usr/include and /usr/lib/lib…

**Comment by theopolis** [comment]
The cause of the linking error in osquery comes from: https://github.com/facebook/osquery/blob/master/CMakeLists.txt#L280

The following logic:
```cmake
# Make sure the OpenSSL/ssl library has a TLS client method.
# Prefer the highest version of TLS, but accept 1.2, 1.1, or 1.0.
include(CheckLibraryExists)
CHECK_LIBRARY_EXISTS(${OPENSSL_SSL_LIBRARY} "TLSv1_2_client_method" "" OPENSSL_TLSV12)
CHECK_LIBRARY_EXISTS(${OPENSSL_SSL_LIBRARY} "TLSv1_1_client_method" "" OPENSSL_TLSV11)
CHECK_LIBRARY_EXISTS(${OPENSSL_SSL_LIBRARY} "TLSv1_client_method" "" OPENSSL_TLSV10)

# Add a define based on the highest TLS version found. Fatal if no TLS client.
if(OPENSSL_TLSV12)
  add_definitions(-DSSL_TXT_TLSV1_2)
elseif(OPENSSL_TLSV11)
  add_definitions(-DSSL_TXT_TLSV1_1)
elseif(OPENSSL_TLSV10)
  add_definitions(-DSSL_TXT_TLSV1)
else()
  message(FATAL "Cannot find any TLS client methods")
endif()
```

So the "best" TLS client is selected. On my system `/usr/include/openssl/ssl.h`:
```
$ grep "TLS.*client" /usr/include/openssl/ssl.h 
/* Use Cisco's "speshul" version of DTLS_BAD_VER (as client)  */
SSL_METHOD *TLSv1_client_method(void) DEPRECATED_IN_MAC_OS_X_VERSION_10_7_AND_LATER;	/* TLSv1.0 */
SSL_METHOD *DTLSv1_client_method(void) DEPRECATED_IN_MAC_OS_X_VERSION_10_7_AND_LATER;	/* DTLSv1.0 */
```

I'm guessing your `ssl.h` contains the TLS1.1/1.2 method defines.

**Comment by umareddy** [comment]
Yes my ssl.h file contains TSL1.1/1.2 methods:
grep "TLS.*client" /opt/local/include/openssl/ssl.h
/* Use Cisco's "speshul" version of DTLS_BAD_VER (as client)  */
const SSL_METHOD *TLSv1_client_method(void);	/* TLSv1.0 */
const SSL_METHOD *TLSv1_1_client_method(void);	/* TLSv1.1 */
const SSL_METHOD *TLSv1_2_client_method(void);	/* TLSv1.2 */
const SSL_METHOD *DTLSv1_client_method(void);	/* DTLSv1.0 */

Thanks for your help, I will close the issue I raised.

**Comment by Motti-Shneor** [comment]
Guys, although I come from a totally different direction - pure Xcode (no external build system, no Brew, no MacPorts whatsoever) I experienced a similar problem. The important clue I have to offer is that: Xcode ignores specifically-pated-libraries, and when linking will ALWAYS use its standard library search paths - thus finding the old, incompatible versions of libcrypto UNLESS you add specific library search path to the directory containing the specific version you want to link against.

**Comment by guerrillera** [comment]
No entiendo nada, viene en inglés y aparte no me entero , ayúdame 

Enviado desde mi Windows Phone

-----Mensaje original-----
De: "Motti Shneor" <notifications@github.com>
Enviado: ‎25/‎01/‎2016 17:33
Para: "facebook/osquery" <osquery@noreply.github.com>
Asunto: Re: [osquery] Error linking: missing OpenSSL TLS1.1/1.2 symbols (#1373)

Guys, although I come from a totally different direction - pure Xcode (no external build system, no Brew, no MacPorts whatsoever) I experienced a similar problem. The important clue I have to offer is that: Xcode ignores specifically-pated-libraries, and when linking will ALWAYS use its standard library search paths - thus finding the old, incompatible versions of libcrypto UNLESS you add specific library search path to the directory containing the specific version you want to link against.

**Comment by theopolis** [comment]
Hey @Motti-Shneor, are you using CMake to generate the Xcode project?

**Comment by Motti-Shneor** [comment]
Nope. This is a real Mac development project :) started on Xcode, using the UI for adding sources and resources, etc. I prebuilt openssl 1.0.2e outside of the project using the build-scripts supplied, than dragged the actual products (libcrypto.a and libssl.a) into the project, and added them to my target. I verified in the build-phases tab that they were indeed in the link-with-libraries list, and tried to build my project. 

I experienced strange missing-symbols link errors, and last I understood that the actual libcrypto.a  Xcode was using came from the OS (or maybe the SDK), but not the ones I manually dragged into my prtoject. The only cure I found was to add the library-search-path for my prebuilt openssl 1.0.2 libs on disk (which unfortunately dirties my git repository, because they're outside the repo). I'll have to simply copy those libraries into my project's directory and push them as binaries (which I hate to do).
